### PR TITLE
CNN encoder and AEF data

### PR DESCRIPTION
**Pull request by vdplasthijs**
## What does this PR do?

- Support AEF data (in `ButterflyDataset`). Because the AEF dataset misses a few data points, this requires resizing of `self.records` and `self._len` (for the trainer). This in turn requires calling `setup()` in `__init__()`, and therefore I moved the `pooch` import to an if-statement to not break the tests and logged an issue #23 
- Implemented (Resnet) CNN EO encoder. 
- Updated configs and tests accordingly.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?

### CODE REVIEW

File: src/models/components/eo_encoders/cnn_encoder.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        eo_data = batch.get("eo", {})
+        assert self.eo_data_name in eo_data, f"eo['{self.eo_data_name}'] not found in batch"
+        # assert not torch.any(torch.isnan(eo_data[self.eo_data_name])), f"EO data for modality {self.eo_data_name} contains NaNs in the batch."
+        feats = self.eo_encoder(eo_data[self.eo_data_name])
+        n_nans = torch.sum(torch.isnan(feats)).item()
+        assert (
+            n_nans == 0
+        ), f"CNNEncoder output contains {n_nans}/{feats.numel()} NaNs PRIOR to normalization with data min {eo_data[self.eo_data_name].min()} and max {eo_data[self.eo_data_name].max()}."
+        if self.output_normalization == "l2":
```
Comment (gabrieletijunaityte) [inline src/models/components/eo_encoders/cnn_encoder.py]: Isn't normalisation bound by loss? Right now, normalisation is also in the CLIP loss. So we need to separate this and, for consistency, place it either in the model or the loss.

File: src/models/components/eo_encoders/cnn_encoder.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    def set_n_input_bands(self, n_bands: int | None = None) -> None:
+        """Sets number of input bands based on eo_data_name if n_bands is None.
+
+        :param n_bands: number of input bands
+        :return: None
+        """
+        if n_bands is None:  # infer from eo_data_name
+            if self.eo_data_name == "s2":
+                self.input_n_bands = 4
```
[the remark this answers is not in the document]
Comment (vdplasthijs) [reply src/models/components/eo_encoders/cnn_encoder.py]: Yes good point! I'll change this

[the remark this answers is not in the document]
Comment (gabrieletijunaityte) [reply src/models/components/eo_encoders/cnn_encoder.py]: Do you mean we would have two eo encoders? Otherwise, I would ensure the data matches the encoder through configurations (and could include an assertion check).

**Review by gabrieletijunaityte — APPROVED** [review]
